File: HW1/HW1CodesSolution/HW1functions.py

import numpy as np

def StiffMat(Con,Area,E,length):    # define stiffness matrix
	K = np.zeros([len(length)+1,len(length)+1]) # global stiffness matrix of correct size
	a = 0
	while a < len(length): # loop over the elements 
		k = E*Area[a]/length[a] # calculate equivalent stiffness 
		Kelem = ElemStiffMat(k)
		# on following lines the parts of the element stiffness matrix are placed in the global stiffness matrix
		K[Con[a,0]-1,Con[a,0]-1] = K[Con[a,0]-1,Con[a,0]-1] + Kelem[0,0]
		K[Con[a,1]-1,Con[a,1]-1] = K[Con[a,1]-1,Con[a,1]-1] + Kelem[1,1]
		K[Con[a,0]-1,Con[a,1]-1] = K[Con[a,0]-1,Con[a,1]-1] + Kelem[0,1]
		K[Con[a,1]-1,Con[a,0]-1] = K[Con[a,1]-1,Con[a,0]-1] + Kelem[1,0]
		a = a+1
	return K

def ElemStiffMat(k):
	Kelem = np.zeros([2,2]) # define element stiffness matrix of correct size
    # on following lines the stiffness elements are placed on the correct location
	Kelem[0,0] = k
	Kelem[1,1] = k
	Kelem[0,1] = -k
	Kelem[1,0] = -k
	return Kelem
	
def ApplyBC(F,BC):
	keep = np.arange(len(F)) # initially all nodes are to be kept 
	nodes = BC[0,:] # fixed nodes are defined in first part of BC
	ind =np.zeros(len(nodes))
	a = 0
	while a < len(nodes):
		ind[a] = np.where(keep == nodes[a]-1)[0][0] # check whether node has a boundary condition applied or not 
		a = a + 1
	keep = (set(keep) - set(ind))
	keep = list(keep)
	
	return keep 
	
def CalFred(F,keep,BC,K):
	Freduced = F[keep] # initially the reduced force vector is equal to the forces applied at these nodes
	nodes = BC[0,:] # the nodes that are fixed have to be added (if displacement is non-zero)
	nodes = nodes.astype(int) # make sure it are all integers
	nodebc = 0
	while nodebc < len(nodes):
		a = 0
		while a < len(Freduced):
			Freduced[a] = Freduced[a] - BC[1,nodebc] * K[keep[a],nodes[nodebc]-1] # subtract what was removed from Left-hand side to the force vector
			
			a = a + 1
		nodebc = nodebc + 1
	
	return Freduced
	
def calcStrain(u,length): # calculate strain for each element 
	strain = np.zeros(len(length)) # calculate vector of right size 
	a = 0
	while a < len(strain):
		strain[a] = (u[a+1]-u[a])/length[a] # calculate strain of an element 
		a = a + 1
	return strain
	
def calcStress(strain, E): # calculate stress for each element 
	stress = strain * E
	return stress 

def DefineConn(Nelem): # define connectivity matrix 
	Con = np.zeros([Nelem,2]) # create vector of right size
	a = 0
	while a < Nelem: # fill up the connectivity matrix for this specific case 
		Con[a,0] = a+1 
		Con[a,1] = a+2
		a = a + 1
	Con = Con.astype(int) # make sure it are all integers
	return Con

def DefForceVec(Nelem,F): # define force vector (force on last node)
	Fvec = np.zeros(Nelem+1) # vector of right size, all zeros 
	Fvec[-1]=F # put the force on the final node 
	return Fvec

def CalcNodePos(Nelem,L):
	NodePos = np.zeros(Nelem+1) # create vector of correct size 
	a = 0
	while a < Nelem:
		NodePos[a+1] = NodePos[a] + L/Nelem # fill up the node positions
		a = a + 1
	return NodePos 

def CalcArea(NodePos,L): # function to calculate the area 
	Nelem = len(NodePos) - 1
	Area = np.zeros(Nelem) # create vector of right size
	w1 = 50
	w2 = 25
	t = 3.125
	a = 0
	while a < Nelem:
		Area[a] = (w1 + (w2-w1)/L*(NodePos[a]+NodePos[a+1])/2)*t #area calculated in centre of element 
		a = a + 1
	return Area
	
def CalcLength(L,Nelem):
	Length = np.zeros(Nelem) # create vector of right size 
	a = 0
	while a < Nelem:
		Length[a] = L/Nelem # put length of element at right position 
		a = a + 1
	return Length

def calcDisp(L, Nelem, F, E, BC): # main function called from other program 
	F = DefForceVec(Nelem,F) # calculate force vector in function
	Con = DefineConn(Nelem) # calculate connectivity matrix in function
	NodePos = CalcNodePos(Nelem,L) # calculate Node position vector in function
	Area = CalcArea(NodePos,L) # calculate area in function
	length = CalcLength(L,Nelem) # calculate length in function
	
	K = StiffMat(Con,Area,E,length) # calculate global stiffness matrix  in function
	
	keep = ApplyBC(F,BC) #determine which indices to keep 
	
	KredInt = K[keep,:] # only keep the required rows
	Kreduced = KredInt[:,keep] # only keep required columns 
	
	Freduced = CalFred(F,keep,BC,K) # calculate reduced force vector 
	
	u = np.linalg.solve(Kreduced, Freduced) # solve for unknown displacements
	
	utotal = np.zeros(len(F)) # vector of right size for displacement 
	utotal[keep] = u # put calculated displacement on right locations
	utotal[BC[0,:]-1] = BC[1,:] # insert the applied BC
	
	strain = calcStrain(utotal,length) # calculate strain in function
	stress = calcStress(strain, E) # calculate stress in function
	
	return utotal, strain, stress 

File: HW1/HW1CodesSolution/HW1functions.m
function [uTotal, strain, stress] = HW1functions(L, Nelem, F, E, BC)

Fvec = zeros(Nelem+1,1); % create force vector of right size
Fvec(end) = F; % put force on final node

NodePos = 0:L/Nelem:L; % calculate node positions

Con = [1:Nelem;2:Nelem+1]'; % define connectivity matrix for this case

Area = CalcArea(NodePos,L); % call function for area calculation

len = NodePos(2:end)-NodePos(1:end-1); % calculated length of each element

K = StiffMat(Con,Area,E,len); % call function to build global stiffness matrix

remov = BC(1,:); % these are the degrees of freedom (nodes) that need to be removed
keep = setdiff(1:length(NodePos),remov); % all other degrees of freedom remain 

Kred = K(keep,keep); % reduced stiffness matrix with only free degrees of freedom: the ones linked to boundary conditions have been removed

Freduced = CalFred(Fvec,keep,BC,K); % calculate the reduced force vector in a function

u = Kred\Freduced; % solve for unknown displacements
uTotal = zeros(1,length(NodePos)); % create vector of the right size
uTotal(keep) = u; % put calculated displacement on right locations
uTotal(BC(1,:)) = BC(2,:); % insert the applied BC

strain = calcStrain(uTotal,len); % calculate strain in function
stress = calcStress(E,strain); % calculate stress in function

end

function Area = CalcArea(NodePos,L)
    w1 = 50;
	w2 = 25;
	t = 3.125;
    Area = (w1 + (w2-w1)/L*(NodePos(1:end-1)+NodePos(2:end))/2)*t; % area calculated in the middle, entire vector in one go
end

function K = StiffMat(Con,Area,E,len)
    K = zeros(length(len)+1); % create global stiffness matrix
    for a = 1:length(len)
        k = E*Area(a)/len(a); % calculate equivalent stiffness of elements
        Kelem = ElemStiffMat(k);
        % add the element stiffness matrix at the correct location of the elements using the connectivity matrix 
		K(Con(a,1),Con(a,1)) = K(Con(a,1),Con(a,1)) + Kelem(1,1);
		K(Con(a,2),Con(a,2)) = K(Con(a,2),Con(a,2)) + Kelem(2,2);
		K(Con(a,1),Con(a,2)) = K(Con(a,1),Con(a,2)) + Kelem(1,2);
		K(Con(a,2),Con(a,1)) = K(Con(a,2),Con(a,1)) + Kelem(2,1);
    end
end

function Kelem = ElemStiffMat(k)
    Kelem = zeros(2,2); % define element stiffness matrix of correct size
    % on following lines the stiffness elements are placed on the correct
    % location.
    Kelem(1,1) = k;
    Kelem(2,2) = k;
    Kelem(1,2) = -k;
    Kelem(2,1) = -k;
end

function Fred = CalFred(F,keep,BC,K)

    Fred = F(keep); % initially the reduced force vector is equal to the forces applied at these nodes
    nodes = BC(1,:); % the nodes that are fixed have to be added (if displacement is non-zero)
    for i = 1:length(nodes)
        for i2 = 1:length(Fred)
            Fred(i2) = Fred(i2) - BC(2,i) * K(keep(i2),nodes(i)); % subtract what was removed from Left-hand side to the force vector
        end
    end
end

function strain = calcStrain(u,len)
    strain = (u(2:end)-u(1:end-1))./len;
end

function stress = calcStress(E,strain)
    stress = E*strain;
end